server/src/main/java/mineclone/server/world/ServerWorld.java

```diff
@@ -125,164 +133,206 @@ public void growTree(IBlockPosition pos) {
-	@Override
-	public void updateNeighbor(IBlockPosition pos, Direction fromDir, IBlockState fromState, int updateFlags) {
-		if ((updateFlags & STATE_UPDATE_FLAG) != 0)
-			dispatchStateUpdate(pos, fromDir, fromState);
-
-		if ((updateFlags & BLOCK_UPDATE_FLAG) != 0)
-			dispatchBlockUpdate(pos, fromDir, fromState);
 
-		if ((updateFlags & INVENTORY_UPDATE_FLAG) != 0)
-			dispatchInventoryUpdate(pos, fromDir, fromState);
+	private void tryUpdateNeighborShapes(IBlockPosition pos, IBlockState state, int flags) {
+		if ((flags & SetBlockFlags.UPDATE_NEIGHBOR_SHAPES) != 0) {
+			state.updateNeighborShapes(this, pos);
+		}
 	}
-	
-	private void dispatchStateUpdate(IBlockPosition pos, Direction fromDir, IBlockState fromState) {
-		getBlockState(pos).onStateUpdate(this, pos, fromDir, fromState);
+
+	@Override
+	public void updateNeighborsExceptFrom(IBlockPosition pos, Direction exceptDir) {
+		for (Direction dir : Direction.ALL) {
+			if (dir != exceptDir) {
+				updateNeighbor(pos.offset(dir));
+			}
+		}
 	}
 
-	private void dispatchBlockUpdate(IBlockPosition pos, Direction fromDir, IBlockState fromState) {
-		getBlockState(pos).onBlockUpdate(this, pos, fromDir, fromState);
+	@Override
+	public void updateNeighbor(IBlockPosition pos) {
+		getBlockState(pos).update(this, pos);
 	}
-	
-	private void dispatchInventoryUpdate(IBlockPosition pos, Direction fromDir, IBlockState fromState) {
-		getBlockState(pos).onInventoryUpdate(this, pos, fromDir, fromState);
+
+	@Override
+	public void updateNeighborShapesExceptFrom(IBlockPosition pos, IBlockState state, Direction exceptDir) {
+		for (Direction dir : Direction.ALL) {
+			if (dir != exceptDir) {
+				updateNeighborShape(pos.offset(dir), dir.getOpposite(), pos, state);
+			}
+		}
 	}
-	
+
 	@Override
-	public int getPower(IBlockPosition pos, int powerFlags) {
-		return getPowerExceptFrom(pos, null, powerFlags);
+	public void updateNeighborShape(IBlockPosition pos, Direction dir, IBlockPosition neighborPos, IBlockState neighborState) {
+		getBlockState(pos).updateShape(this, pos, dir, neighborPos, neighborState);
 	}
 
 	@Override
-	public int getPowerExceptFrom(IBlockPosition pos, Direction exceptDir, int powerFlags) {
+	public int getSignalExceptFrom(IBlockPosition pos, Direction exceptDir, SignalType type) {
 		int highestPower = 0;
-		
-		for (Direction dir : Direction.DIRECTIONS) {
+
+		for (Direction dir : Direction.ALL) {
 			if (dir != exceptDir) {
-				int power = getPowerFrom(pos, dir, powerFlags);
-				
+				int power = getSignalFrom(pos, dir, type);
+
 				if (power > highestPower)
 					highestPower = power;
 			}
 		}
-		
+
 		return highestPower;
 	}
-	
+
+	@Override
+	public int getSignalFrom(IBlockPosition pos, Direction dir, SignalType type) {
+		Direction opp = dir.getOpposite();
+		IBlockPosition neighborPos = pos.offset(dir);
+		IBlockState neighborState = getBlockState(neighborPos);
+
+		int signal = neighborState.getSignal(this, neighborPos, opp, type);
+
+		if (signal > type.max()) {
```

Comment:
Should be greater than _or equal_. Does not make sense to check further if the signal is exactly max value.
```suggestion
		if (signal >= type.max()) {
```
Same goes for multiple occurrences below.

Reply:
fixed!